codegen-core/src/main/kotlin/software/amazon/smithy/rust/codegen/core/smithy/generators/protocol/ProtocolTestGenerator.kt

```diff
@@ -0,0 +1,350 @@
+            "require_headers",
+        )
+    }
+
+    fun checkForbidHeaders(
+        rustWriter: RustWriter,
+        actualExpression: String,
+        forbidHeaders: List<String>,
+    ) {
+        basicCheck(
+            forbidHeaders,
+            rustWriter,
+            "forbidden_headers",
+            actualExpression,
+            "forbid_headers",
+        )
+    }
+
+    fun checkHeaders(
+        rustWriter: RustWriter,
+        actualExpression: String,
+        headers: Map<String, String>,
+    ) {
+        if (headers.isEmpty()) {
+            return
+        }
+        val variableName = "expected_headers"
+        rustWriter.withBlock("let $variableName = [", "];") {
+            writeWithNoFormatting(
+                headers.entries.joinToString(",") {
+                    "(${it.key.dq()}, ${it.value.dq()})"
+                },
+            )
+        }
+        assertOk(rustWriter) {
+            write(
+                "#T($actualExpression, $variableName)",
+                RuntimeType.protocolTest(codegenContext.runtimeConfig, "validate_headers"),
+            )
+        }
+    }
+
+    fun basicCheck(
+        params: List<String>,
+        rustWriter: RustWriter,
+        expectedVariableName: String,
+        actualExpression: String,
+        checkFunction: String,
+    ) {
+        if (params.isEmpty()) {
+            return
+        }
+        rustWriter.withBlock("let $expectedVariableName = ", ";") {
+            strSlice(this, params)
+        }
+        assertOk(rustWriter) {
+            rustWriter.rust(
+                "#T($actualExpression, $expectedVariableName)",
+                RuntimeType.protocolTest(codegenContext.runtimeConfig, checkFunction),
+            )
+        }
+    }
+
+    /**
+     * Wraps `inner` in a call to `aws_smithy_protocol_test::assert_ok`, a convenience wrapper
+     * for pretty printing protocol test helper results.
+     */
+    fun assertOk(
+        rustWriter: RustWriter,
+        inner: Writable,
+    ) {
+        rustWriter.rust("#T(", RuntimeType.protocolTest(codegenContext.runtimeConfig, "assert_ok"))
+        inner(rustWriter)
+        rustWriter.write(");")
+    }
+
+    private fun strSlice(
+        writer: RustWriter,
+        args: List<String>,
+    ) {
+        writer.withBlock("&[", "]") {
+            rust(args.joinToString(",") { it.dq() })
+        }
+    }
+}
+
+/**
+ * Service shape IDs in common protocol test suites defined upstream.
+ */
+object ServiceShapeId {
+    const val AWS_JSON_10 = "aws.protocoltests.json10#JsonRpc10"
+    const val AWS_JSON_11 = "aws.protocoltests.json#JsonProtocol"
+    const val REST_JSON = "aws.protocoltests.restjson#RestJson"
+    const val REST_XML = "aws.protocoltests.restxml#RestXml"
+    const val AWS_QUERY = "aws.protocoltests.query#AwsQuery"
+    const val EC2_QUERY = "aws.protocoltests.ec2#AwsEc2"
+    const val REST_JSON_VALIDATION = "aws.protocoltests.restjson.validation#RestJsonValidation"
+}
```

Comment:
Don't we need one for CBOR?

Reply:
I assume it'll be added in the CBOR PR once this PR has been merged to main and then the CBOR PR merges the latest main into it (since this PR is refactoring w.r.t the main branch, it's probably supposed to not know about CBOR 🙂)

Reply:
Actually, some of these are unused. We only need to refer to service shapes containing failing test cases. I've removed the unused ones.